lib/src/components/NxToast/NxToast.tsx

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2019-present Sonatype, Inc.
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/.
+ */
+import React, { useRef, useContext, useEffect, useState } from 'react';
+import classnames from 'classnames';
+
+import { nxToastPropTypes, NxToastProps } from './types';
+import NxCloseButton from '../NxCloseButton/NxCloseButton';
+import NxFontAwesomeIcon from '../NxFontAwesomeIcon/NxFontAwesomeIcon';
+
+import ToastContext from './contexts';
+import { toastTypeMap } from './toastTypeMapping';
+
+const NxToast = (props: NxToastProps) => {
+  const { toastId, className, message, type, toastContainerRef, ...otherProps } = props,
+      toastClass = toastTypeMap[type].class,
+      toastIcon = toastTypeMap[type].icon,
+      toastIconLabel = toastTypeMap[type].iconLabel,
+      [animate, setAnimate] = useState(false),
+      classes = classnames('nx-toast', 'animate', className, toastClass,
+          {'slide-in': animate},
+          {'slide-out': !animate}
+      ),
+      toastContext = useContext(ToastContext),
+      closeBtnRef = useRef<HTMLButtonElement>(null);
+
+  //Fire effect to focus on the close button of the toast that is rendered
+  useEffect(() => {
+    const closeBtn = closeBtnRef.current;
+    if (closeBtn) {
+      closeBtn.focus();
+    }
+
+    //When component is mounted, add class "animate slide-in" to trigger animation
+    setAnimate(true);
+  }, []);
+
+  //Effect to shift focus to the next toast's close button
+  //Dependency on the global toasts array
+  useEffect(() => {
+    if (toastContainerRef) {
+      //Gets the first close button of the child from the parent toast container div
+      const closeBtn = toastContainerRef.current?.querySelectorAll('.nx-btn--close')[0] as HTMLButtonElement;
+      closeBtn.focus();
+    }
+  }, [toastContext?.toasts]);
+
+  const handleCloseClick = () => {
+    setAnimate(false);
+    //Wait for slide-out animation before removing toast from DOM
+    setTimeout(() => {
+      toastContext?.removeToast(toastId);
+    }, 250);
+  };
+
+  return (
+    <div role="alert" { ...otherProps } className={classes} aria-atomic={true}>
```

Comment:
Rather than basically reimplement `NxAlert` here, can we use actual `NxAlert` and just wrap it?

Reply:
> Rather than basically reimplement NxAlert here, can we use actual NxAlert and just wrap it?

Expanding on this, I'd really like to see us avoid reiterating/reimplementing so many of the `NxAlert` concepts in this toast work.  I suspect we could make something where the calling code is expected to pass us an `NxAlert` which we then render in a toast layout.

Reply:
Hey Ross, 
So both `NxToast` and `NxAlert` have success, error, warning and info states. So I was thinking of reusing `NxAlert` and its variations for `NxToast`, and cutting down lines of code by [choosing the correct component variation at runtime](https://reactjs.org/docs/jsx-in-depth.html#choosing-the-type-at-runtime). I was thinking of doing something like this:
```
const alerts = {
  success: NxSuccessAlert,
  error: NxErrorAlert,
  info: NxInfoAlert,
  warning: NxWarningAlert
};

type alertType = keyof typeof alerts;
const NxToast = (type:alertType, message:string) => {
  const NxAlertType = alerts[type]; // 
  return <NxAlertType>{message}</NxAlertType>;
}

// Example
NxToast('info', 'Example text');
```

I greatly simplified the code so the example is easier to read. What do you think of this approach?

Reply:
> I greatly simplified the code so its easier to read. What do you think of this approach?

What I was picturing was more something that would give us an API like
```
<NxToast>
  <NxErrorAlert>...</NxErrorAlert>
</NxToast>
```

Thoughts?

Reply:
> What I was picturing was more something that would give us an API like
> 
> ```
> <NxToast>
>   <NxErrorAlert>...</NxErrorAlert>
> </NxToast>
> ```
> 
> Thoughts?

I like your proposal, but I'm having trouble implementing it. I think what you're proposing is that when `NxToast` is rendered, it should have exactly one `NxAlert`, `NxErrorAlert`, etc... as its child. If that's true, what's your recommendation on typechecking `NxToast`'s child and confirming its the same type as `NxAlert` or its variations? A Google search did not give any straightforward solutions.
```
<NxToast>
  <NxErrorAlert>...</NxErrorAlert>
  <NxSucessAlert>...</NxSuccessAlert>
</NxToast>

<NxToast>
  <div class="alert">...</div>
</NxToast>
```

Also, what do you think of this as an alternative?
```
<NxToast alert={NxErrorAlert}>Test Message</NxToast>
```

Reply:
> If that's true, what's your recommendation on typechecking NxToast's child and confirming its the same type as NxAlert or its variations?

I've tried to do that sort of thing in the past and found it impossible.  The typings of react would have to be adjusted for that to work.  The closest we can get is to enforce that `children` is a `ReactElement` rather than `ReactNode`.  The rest just has to be documentation.

> <NxToast alert={NxErrorAlert}>Test Message</NxToast>

Probably not worth it if the only thing it helps with is typechecking.  We want the API to feel natural.

Reply:
> What I was picturing was more something that would give us an API like
> 
> ```
> <NxToast>
>   <NxErrorAlert>...</NxErrorAlert>
> </NxToast>
> ```

As discussed in our meeting, the changes inside the gallery to render the NxAlerts are here: [8ce72d2](https://github.com/sonatype/sonatype-react-shared-components/pull/688/commits/8ce72d2768d7f88a888927cd56c8713111658fb0)

(pardon some of the repetitive other type definitions , this commit was a work in progress at the time)